### ISSUE TITLE
Allow prior to be a callable in random module

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#1224]
Currently, [random_flax_module](http://num.pyro.ai/en/stable/primitives.html#numpyro.contrib.module.random_flax_module) and [random_haiku_module](http://num.pyro.ai/en/stable/primitives.html#numpyro.contrib.module.random_haiku_module) assume that `prior` is either a distribution or a dict of distributions. In case we have a large network and want to specify priors for specific parameters like the weight of the last layer, we need to find corresponding names of those parameters. It is a bit inconvenient. It would be nice to allow a callable `prior` argument like
```
def get_prior(name, shape):
    if "weight_last" in name:
        return dist.Normal(0, 1).expand([shape])
```
where `name` is the parameter full name and `shape` is its corresponding shape.

This is a nice feature and supporting this is not complicated. We just need to add a [check](https://github.com/pyro-ppl/numpyro/blob/d3a4ca116d8d99372fae030fe04990db6a522055/numpyro/contrib/module.py#L226) here such that when `prior` is a callable, we get `d` by calling `prior(flatten_name, param_shape)`. We'll need a few tests for this feature and this would be a nice chance to get familiar with using flax/haiku in numpyro.

**Comment by Carlosbogo** [comment]
Hi, I would like to work on this, but I haven't worked on this repository before. Could you explain which special parameters could be nice to add?

**Comment by fehiepsi** [comment]
You don't need to add new parameters. Just add a check if prior is a callable then we get its output.

**Comment by Carlosbogo** [comment]
Sorry, I meant what could be a list of those "special parameters" you mention, and where I could find which distribution should be associated with each one of them.
You use the example of weight with the normal distribution associated to it, but I'm not sure if there is a list of other parameters that have an associated distribution in this repository.

**Comment by fehiepsi** [comment]
I think we would like to let users implement their own callable `prior`. For example, `prior` can be the following function
```python
def get_prior(name, shape):
    if "bias_arbitrary_on_the_sky" in name:
        return dist.LogNormal(0, 1).expand([shape])
```

**Comment by Carlosbogo** [comment]
I'm not sure if I understand what you mean: Should it be just a placeholder function waiting for each user to decide what parameters are needed to specify or should I make a more general function of the form (not the actual function, just a sketch)
```
def get_prior(name, shape, parameter, distribution, interval):
    if parameter in name:
        return distribution(interval).expand([shape])
```

**Comment by fehiepsi** [comment]
It is just a user-defined function takes inputs are name and shape of the parameters and returns the user-defined distributions with corresponding shape.

In the random flax module utility, we will use that function to set priors to corresponding sites. If the above function returns None, we won't set prior. Otherwise, we will consider its output as prior.

**Comment by fehiepsi** [comment]
This is resolved in #1227. Thanks, @Carlosbogo!
